### PR TITLE
Added the string formatting lab

### PR DESCRIPTION
**Pull request by SLammertink**
Added the strng formatting lab from lesson 03.

### CODE REVIEW

File: students/SLammertink/Lesson03/strformat_lab.py [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+# Task 1
+''' Write a format string that will take the following four element tuple:
+( 2, 123.4567, 10000, 12345.67)
+and produce:
+'file_002 :   123.46, 1.00e+04, 1.23e+04' '''
+
+StrList = (2, 123.4567, 10000, 12345.67)
+
+def Task1():
+    print(f"file_{StrList[0]:03d} :  {StrList[1]:.2f}, {StrList[2]:.2e}, {StrList[3]:.2e}")
```
Comment (charlesrobison) [inline students/SLammertink/Lesson03/strformat_lab.py]: Good - make sure function and method names are lower case - any with multiple words use underscore to separate.

File: students/SLammertink/Lesson03/strformat_lab.py [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+
+'''Using your results from Task One, repeat the exercise, but this time using
+an alternate type of format string (hint: think about alternative ways to use .format()
+(keywords anyone?), and also consider f-strings if you’ve not used them already).
+'''
+
+
+def Task2():
+    print("file_{0:03d} :  {1:.2f}, {2:.2e}, {3:.2e}".format(StrList[0], StrList[1], StrList[2], StrList[3]))
+    print("file_%03d :  %.2f, %.2e, %.2e" % StrList)
```
Comment (charlesrobison) [inline students/SLammertink/Lesson03/strformat_lab.py]: Nice!  Same comment as above - lower case function names.

File: students/SLammertink/Lesson03/strformat_lab.py [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+
+# Task 3
+
+''' Rewrite: "the 3 numbers are: {:d}, {:d}, {:d}".format(1,2,3) to take an arbitrary number of values.'''
+
+def formatter(in_tuple):
+    formatter_string = "the " + str(len(in_tuple)) + " numbers are " + ", ".join(["{}"]*len(in_tuple)).format(*in_tuple)
+    print(formatter_string)
```
Comment (charlesrobison) [inline students/SLammertink/Lesson03/strformat_lab.py]: Good!

File: students/SLammertink/Lesson03/strformat_lab.py [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+
+'''Given a 5 element tuple: ( 4, 30, 2017, 2, 27)
+use string formating to print: '02 27 2017 04 30' '''
+
+five_tuple = ( 4, 30, 2017, 2, 27)
+
+def TaskFour(five_tuple):
+    reformat = "{0:02d} {1:d} {2:d} {3:02d} {4:d}"
+    print(reformat.format(five_tuple[3], five_tu
+                           uit_list = ['oranges', 1.3, 'lemons', 1.1]
```
Comment (charlesrobison) [inline students/SLammertink/Lesson03/strformat_lab.py]: Something odd going on here - typo perhaps?  Doesn't process correctly - five_tu?  uit_list?

File: students/SLammertink/Lesson03/strformat_lab.py [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+print(f'The weight of an {str(fruit_list[0])[:-1].upper()} is {fruit_list[1] * 1.2} and the weight of a {str(fruit_list[2].upper())[:-1]} is {fruit_list[3] * 1.2}')
+
+# Task 6
+
+# Write some Python code to print a table of several rows, each with a name, an age and a cost. Make sure some of the costs are in the hundreds and thousands to test your alignment specifiers.
+
+
+rows = [('Name', 'Age', 'Cost'),('Seb', 50, 200),('Kim', 47, 5000), ('Mike', 25, 375.25)]
+for row in rows:
+    print("{:^10}{:^10}{:^10}".format(*row))
```
Comment (charlesrobison) [inline students/SLammertink/Lesson03/strformat_lab.py]: Great!

File: students/SLammertink/Lesson03/strformat_lab.py [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+                           uit_list = ['oranges', 1.3, 'lemons', 1.1]
+
+# Task 5
+
+
+print(f'The weight of an {str(fruit_list[0])[:-1]} is {fruit_list[1]} and the weight of a {str(fruit_list[2])[:-1]} is {fruit_list[3]}')
+
+# Now see if you can change the f-string so that it displays the names of the fruit in upper case, and the weight 20% higher
+
+print(f'The weight of an {str(fruit_list[0])[:-1].upper()} is {fruit_list[1] * 1.2} and the weight of a {str(fruit_list[2].upper())[:-1]} is {fruit_list[3] * 1.2}')
```
Comment (charlesrobison) [inline students/SLammertink/Lesson03/strformat_lab.py]: Works well.

File: students/SLammertink/Lesson03/strformat_lab.py (hunk 4 above)
Comment (SLammertink) [reply students/SLammertink/Lesson03/strformat_lab.py]: Hi must be a typo, that uit_list comes from a  different exercise..strange..it should be:
print(reformat.format(five_tuple[3], five_tuple[4], five_tuple[2], five_tuple[0], five_tuple[1]))

File: students/SLammertink/Lesson03/strformat_lab.py (hunk 1 above)
Comment (SLammertink) [reply students/SLammertink/Lesson03/strformat_lab.py]: ok thanks will change the names to lower case in the next exercises!

**Comment by natasha-aleksandrova** [comment]
@charlesrobison this PR now contains mailroom assignment, can you please review that as well before I merge it.

File: students/SLammertink/Lesson03/mailroom_1.py [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+# The Functions used:
+
+def name_in_list():
+    #Checks whether is name is in the name list if not it adds it
+    input_name = input("Please enter the donor's name: ")
+    if input_name not in name_list:
+        print('')
+        print("That name is not in the donor list, will add it now.") # notify the user that the name will be added to the list
+        print('')
+        name_list.append(input_name)
```
Comment (charlesrobison) [inline students/SLammertink/Lesson03/mailroom_1.py]: Something unexpected happening here.  New donors aren't being added as expected.  Trying to run a report after adding a donor causes an error.  Take a look at exception handling.

File: students/SLammertink/Lesson03/mailroom_1.py [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+# Author: SLammertink
+# UW Self Paced Lesson 03
+# Mailroom part 1
+
+
+# Initiate the lists used
+don_list= [75, 25346.25, 125, 25, 200.50] # list with the donations
+name_list = ['Sukhmani Travers', 'Sebastien Mayo', 'Aryan Davila', 'Zayan Langley', 'Charlotte Bates'] #list with the donor names
+count_list =  [1, 2, 1, 3, 2] # list with a count of the donations given
+choices_list = ['Sending a Thank You', 'Creating a Report', 'quit'] # list from where user can chose
```
Comment (natasha-aleksandrova) [inline students/SLammertink/Lesson03/mailroom_1.py]: your code would be much simpler if you had a single list:
```
donors = [('Aryan Davila', [100, 200]),]
```

File: students/SLammertink/Lesson03/mailroom_1.py [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+                count_list[i] = (count_list[i] + 1)
+                return(don_list, count_list)
+
+# Creating a report function:
+
+def report():
+    print("{:^30}{:5}{:^20}{:5}{:^30}{:5}{:^20}{:5}".format('Name', '|', 'Donation','|', 'Number of gifts','|', 'Average gift', '|'))
+    print("{:-<116}".format(''))
+    print(' ')
+    for i, j, k in zip(name_list, don_list, count_list):
```
Comment (natasha-aleksandrova) [inline students/SLammertink/Lesson03/mailroom_1.py]: this report should be sorted, please check the spec

File: students/SLammertink/Lesson03/mailroom_1.py [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+            print('Here are the donors already in the list: ')
+            print('')
+            print(name_list)
+            print('')
+        elif choice == "1":
+            print('')
+            name_in_list()
+        elif choice == "2":
+                print('')
+                report()
```
Comment (natasha-aleksandrova) [inline students/SLammertink/Lesson03/mailroom_1.py]: over-indented.

File: students/SLammertink/Lesson03/mailroom_1.py [hunk 11]
```diff
@@ -0,0 +1,82 @@
...
+            print(i + 1, j)
+        choice = input("Please chose an option, to see the list of donors type 'list': ")
+        if choice == 'list':
+            print('')
+            print('Here are the donors already in the list: ')
+            print('')
+            print(name_list)
+            print('')
+        elif choice == "1":
+            print('')
```
Comment (natasha-aleksandrova) [inline students/SLammertink/Lesson03/mailroom_1.py]: the print statements here are making the entry point a little "messy"; in next version you'll see that they will have to move.

File: students/SLammertink/Lesson03/mailroom_1.py [hunk 12]
```diff
@@ -0,0 +1,82 @@
...
+        input_don = float(input("How much is the donation: "))
+        print('')
+        print("Sending email ----->") # pretending to send the email
+        print("Thank you {} for your generous donation!".format(input_name)) # format the email message
+        print('')
+        for i, name in enumerate(name_list):
+            if name == input_name:
+                don_list[i] = (don_list[i] + input_don)
+                count_list[i] = (count_list[i] + 1)
+                return(don_list, count_list)
```
Comment (natasha-aleksandrova) [inline students/SLammertink/Lesson03/mailroom_1.py]: why is there a return statement here? returned data is not being consumed.

File: students/SLammertink/Lesson03/mailroom_1.py [hunk 13]
```diff
@@ -0,0 +1,82 @@
...
+
+# Initiate the lists used
+don_list= [75, 25346.25, 125, 25, 200.50] # list with the donations
+name_list = ['Sukhmani Travers', 'Sebastien Mayo', 'Aryan Davila', 'Zayan Langley', 'Charlotte Bates'] #list with the donor names
+count_list =  [1, 2, 1, 3, 2] # list with a count of the donations given
+choices_list = ['Sending a Thank You', 'Creating a Report', 'quit'] # list from where user can chose
+
+# The Functions used:
+
+def name_in_list():
```
Comment (natasha-aleksandrova) [inline students/SLammertink/Lesson03/mailroom_1.py]: should this function be named `send_thank_you`?

**Pull request merged**